File: T1.SqlSharp/T1.SqlSharp/Expressions/SqlUnionSelect.cs
using T1.Standard.IO;

namespace T1.SqlSharp.Expressions;

public class SqlUnionSelect : ISqlExpression
{
    public SqlType SqlType => SqlType.UnionSelect;
    public TextSpan Span { get; set; } = new();
    public void Accept(SqlVisitor visitor)
    {
        visitor.Visit_UnionSelect(this);
    }

    public SqlSetOperator Operator { get; set; } = SqlSetOperator.Union;
    public bool IsAll => Operator == SqlSetOperator.UnionAll;
    public required ISqlExpression SelectStatement { get; set; }

    public string ToSql()
    {
        var sql = new IndentStringBuilder();
        sql.Write($"{GetOperatorKeyword()} ");
        sql.Write(SelectStatement.ToSql());
        return sql.ToString();
    }

    private string GetOperatorKeyword()
    {
        return Operator switch
        {
            SqlSetOperator.UnionAll => "UNION ALL",
            SqlSetOperator.Intersect => "INTERSECT",
            SqlSetOperator.Except => "EXCEPT",
            _ => "UNION"
        };
    }
}

File: VimSharp/chrome-extension-sharp/src/t1-sqlts/Expressions/SqlUnionSelect.ts
import { SqlType } from './SqlType';
import { TextSpan } from '../StringParser';
import { ISqlExpression, SqlVisitor } from './ISqlExpression';

export class SqlUnionSelect implements ISqlExpression {
    SqlType: SqlType = SqlType.UnionSelect;
    Span: TextSpan = new TextSpan();
    IsAll: boolean = false;
    SelectStatement!: ISqlExpression;

    Accept(visitor: SqlVisitor): void {
        visitor.Visit_UnionSelect(this);
    }

    ToSql(): string {
        let sql = 'UNION ';
        if (this.IsAll) {
            sql += 'ALL ';
        }
        sql += this.SelectStatement.ToSql();
        return sql;
    }
} 